go-controller/pkg/ovn/dns_name_resolver/external_dns.go

```diff
@@ -0,0 +1,235 @@
+	// that such objects are removed from the cluster, however, the
+	// node will still receive the events related to the unwarranted
+	// object. Any event related to such objects should be ignored.
+	dnsName := string(obj.Spec.Name)
+	objName, exists := extEgDNS.dnsNameToResolver[dnsName]
+	if exists && objName != name {
+		return nil
+	}
+
+	// Add the detials of the object name and the DNS name to the
+	// objDNSName and the dnsNameObj maps.
+	extEgDNS.resolverToDNSName[name] = dnsName
+	extEgDNS.dnsNameToResolver[dnsName] = name
+
+	// Get the addresses corresponding to the DNS name and add them to
+	// the address set corresponding to the DNS name.
+	addresses := []string{}
+	for _, resolvedName := range obj.Status.ResolvedNames {
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			addresses = append(addresses, resolvedAddress.IP)
+		}
+	}
+	err = extEgDNS.dnsTracker.addDNSName(dnsName, addresses)
+	return err
+}
+
+// Add adds the namespace to the set of namespaces where the DNS name is used in the
+// EgressFirewall rules. It also returns the address set corresponding to the DNS name.
+// The address set may be empty at this point if the corresponding DNSNameResolver
+// object's status is still not updated with the associated IP addresses.
+func (extEgDNS *ExternalEgressDNS) Add(namespace, dnsName string) (addressset.AddressSet, error) {
+	return extEgDNS.dnsTracker.addNamespace(namespace, dnsName)
+}
+
+// Delete removes the namespace from the set of namespaces where the DNS name is used in
+// the EgressFirewall rules.
+func (extEgDNS *ExternalEgressDNS) Delete(namespace string) error {
+	return extEgDNS.dnsTracker.deleteNamespace(namespace)
+}
+
+// Run starts the DNSNameResolver controller.
+func (extEgDNS *ExternalEgressDNS) Run() error {
+	return controller.StartControllersWithInitialSync(extEgDNS.syncDNSNames, extEgDNS.controller)
+}
+
+// Shutdown stops the DNSNameResolver controller.
+func (extEgDNS *ExternalEgressDNS) Shutdown() {
+	controller.StopControllers(extEgDNS.controller)
+
+}
+
+// syncDNSNames fetches the DNSNameResolver objects and the EgressFirewall objects to get expected DNS
+// names list, then iterates through the existing address sets to check if the corresponding DNS name
+// exists in the list. If not, then the address set is deleted.
+func (extEgDNS *ExternalEgressDNS) syncDNSNames() error {
+	// Fetch the existing DNSNameResolver objects.
+	dnsNameResolvers, err := extEgDNS.dnsLister.DNSNameResolvers(config.Kubernetes.OVNConfigNamespace).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("syncDNSNames unable to get Egress Firewalls: %w", err)
+	}
+
+	expectedDNSNames := sets.New[string]()
+	for _, dnsNameResolver := range dnsNameResolvers {
+		expectedDNSNames.Insert(string(dnsNameResolver.Spec.Name))
+	}
+
+	// Fetch the existing EgressFirewall objects.
+	egressFirewalls, err := extEgDNS.efLister.List(labels.Everything())
```

Comment:
listing egressfirewalls directly here is racy, because egress firewall event handler is already running, so if an egress firewall is added after you list it, you won't have its dnsName in the expectedDNSNames, and this code will delete a valid address set.
I think if we move this function under `dnsTracker` it will work:
```
func (dnsTracker *dnsTracker) cleanupDB() error {
	dnsTracker.dnsLock.Lock()
	defer dnsTracker.dnsLock.Unlock()

	return dnsTracker.addressSetFactory.ProcessEachAddressSet(dnsTracker.controllerName, ops.AddressSetEgressFirewallDNS,
		func(dbIDs *ops.DbObjectIDs) error {
			addrSetDnsName := dbIDs.GetObjectID(ops.ObjectNameKey)
			if _, ok := dnsTracker.dnsNames[addrSetDnsName]; !ok {

				klog.Infof("Deleting stale address set %s", addrSetDnsName)
				if err := dnsTracker.addressSetFactory.DestroyAddressSet(dbIDs); err != nil {
					klog.Errorf(err.Error())
					return err
				}
			}
			return nil
		})
}
```
locking internal structure here guarantees that no changes are being processed at the same time, and we know that any address sets that are not present in `dnsTracker.dnsNames` may be deleted.
I think we also need to add a comment to the `func (extEgDNS *ExternalEgressDNS) Run()` saying it should be only started after egress firewall handler, and explain cleanup dependencies

Reply:
If we are running the cleanup after every restart, wouldn't the internal map `dnsTracker.dnsNames` be empty at that time? Won't this result in a no-op?

Reply:
right, that is what I meant by "it should be only started after egress firewall handler". A part of egressfirewall handler logic is that it will handle every add event at least once before WatchEgressFirewall returns. but I don't like this logic now, as it will break if we ever switch to level-driven controller for egress firewall.
The general problem is that egress firewall handler should first run its own initialSync to delete all stale egress firewalls. After this is done some dnsName address sets may become stale (not referenced by any ACL) and we need to clean them up. So another idea: we could make `ExternalEgressDNS.DeleteStaleAddrSets` function and call it directly from egressFirewall initialSync https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressfirewall.go#L132 in the very end. The DeleteStaleAddrSets could just look for all unreferenced address sets and delete them, similar to https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/pod_selector_address_set.go#L733

Reply:
Made the suggested changes. Added unit tests as well.